### PR TITLE
daemon: move postSnap and inst.dispatch tests to api_snaps_test.go

### PR DESCRIPTION
**Pull request by pedronis**
Another step cleaning up a bit daemon API tests.

### CODE REVIEW

**Review by bboozzoo — APPROVED** [review]
Thanks

File: daemon/api_snaps_test.go [hunk 1]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+
+		rsp := s.req(c, req, nil).(*daemon.Resp)
+
+		c.Check(rsp.Type, check.Equals, daemon.ResponseTypeError)
+		c.Check(rsp.Status, check.Equals, 400, check.Commentf("%q", action))
+		c.Check(rsp.Result.(*daemon.ErrorResult).Message, check.Equals, expectedErr, check.Commentf("%q", action))
+	}
+}
+
+func (s *snapsSuite) TestPostSnapLeaveCohortRandoAction(c *check.C) {
```
Comment (mvo5) [inline daemon/api_snaps_test.go:1221]: (nitpick) same "Rando"->"Random" typo (or maybe on purpose)?

File: daemon/api_snaps_test.go [hunk 2]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+		errmsg string
+	}
+
+	for i, t := range []T{
+		// TODO: more?
+		{`"cohort-key": "what", "revision": "42"`, `cannot specify both cohort-key and revision`},
+		{`"cohort-key": "what", "leave-cohort": true`, `cannot specify both cohort-key and leave-cohort`},
+	} {
+		buf := strings.NewReader(fmt.Sprintf(`{"action": "refresh", %s}`, t.opts))
+		req, err := http.NewRequest("POST", "/v2/snaps/some-snap", buf)
```
Comment (mvo5) [inline daemon/api_snaps_test.go:1251]: this is so much nicer than the original test that used `s.vars` !

File: daemon/api_snaps_test.go [hunk 3]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+	}
+}
+
+func (s *snapsSuite) TestPostSnapSetsUser(c *check.C) {
+	d := s.daemon(c)
+
+	_, restore := daemon.MockEnsureStateSoon(func(st *state.State) {})
+	defer restore()
+
+	checked := false
```
Comment (mvo5) [inline daemon/api_snaps_test.go:1268]: (nitpick) maybe an int to ensure we called MockSnapstateInstall exactly once?

File: daemon/api_snaps_test.go [hunk 4]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+
+	soon := 0
+	var origEnsureStateSoon func(*state.State)
+	origEnsureStateSoon, restore := daemon.MockEnsureStateSoon(func(st *state.State) {
+		soon++
+		origEnsureStateSoon(st)
+	})
+	defer restore()
+
+	checked := false
```
Comment (mvo5) [inline daemon/api_snaps_test.go:1143]: (nitpick) maybe an int to ensure we called MockSnapstateInstall exactly once?

File: daemon/api_snaps_test.go [hunk 5]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+	rsp := s.req(c, req, user).(*daemon.Resp)
+
+	c.Check(rsp.Type, check.Equals, daemon.ResponseTypeAsync)
+
+	st := d.Overlord().State()
+	st.Lock()
+	defer st.Unlock()
+	chg := st.Change(rsp.Change)
+	c.Assert(chg, check.NotNil)
+	c.Check(checked, check.Equals, true)
```
Comment (mvo5) [inline daemon/api_snaps_test.go:1296]: (nitpick) The original code had a "chg.Summary()" test, maybe this version could do something like
```
c.Check(chg.Tasks()[0].Summary(), Equals, "Doing a fake install")
```
or something similar?

File: daemon/api_snaps_test.go [hunk 6]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+	}
+
+	st := d.Overlord().State()
+	st.Lock()
+	defer st.Unlock()
+	_, _, err := inst.Dispatch()(inst, st)
+	c.Check(err, check.ErrorMatches, "cannot use devmode and jailmode flags together")
+}
+
+func (s *snapsSuite) TestInstallCohort(c *check.C) {
```
Comment (mvo5) [inline daemon/api_snaps_test.go:1428]: Thanks for sorting this in a sensible way, the old file was all over the place

File: daemon/api_snaps_test.go [hunk 7]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+	s.daemonWithFakeSnapManager(c)
+
+	var buf bytes.Buffer
+	buf.WriteString(`{"action": "install"}`)
+	req, err := http.NewRequest("POST", "/v2/snaps/some-snap", &buf)
+	req.RemoteAddr = "pid=100;uid=0;socket=;"
+	c.Assert(err, check.IsNil)
+	req.Header.Add("User-Agent", "some-agent/1.0")
+
+	s.vars = map[string]string{"name": "some-snap"}
```
Comment (mvo5) [inline daemon/api_snaps_test.go:1551]: (nipick) Is this needed? It seems the request does that for the relevant pieces, no?

**Review by mvo5 — APPROVED** [review]
Looks great, thanks for doing this work. I had some nitpicks during the review, feel free to ignore.

File: daemon/api_snaps_test.go [hunk 8]
```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
...
+	c.Assert(err, check.IsNil)
+
+	rsp := s.req(c, req, nil).(*daemon.Resp)
+
+	c.Check(rsp.Type, check.Equals, daemon.ResponseTypeError)
+	c.Check(rsp.Status, check.Equals, 400)
+	c.Check(rsp.Result.(*daemon.ErrorResult).Message, testutil.Contains, `cannot install "ubuntu-core", please use "core" instead`)
+}
+
+func (s *snapsSuite) TestPostSnapCohortRandoAction(c *check.C) {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply daemon/api_snaps_test.go:1204]: I can change it but that was the original naming

[the remark this answers is not in the document]
Comment (pedronis) [reply daemon/api_snaps_test.go:1204]: this is about actions that do not support passing a cohort, I'm s/Rando/Unsupported/

File: daemon/api_snaps_test.go (hunk 1 above)
Comment (pedronis) [reply daemon/api_snaps_test.go:1221]: same

**Pull request merged**